rpc-broker/src/msg.c

```diff
@@ -0,0 +1,245 @@
+/*
+ Copyright (c) 2018 AIS, Inc.
+
+ This program is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+#include "rpc-broker.h"
+
+
+/*
+ * All requests are handled by this function whether they are raw requests
+ * coming from port 5555 or Websocket requests from 8080.  For every policy
+ * rule listed either in /etc/rpc-broker.policy file or the domain-specific
+ * rules listed in xenclient database are passed to `filter` to determine
+ * whether or not to be dropped or passed through.
+ */
+int broker(struct dbus_message *dmsg, int domid)
+{
+    if (!dmsg) {
+        DBUS_BROKER_WARNING("Invalid args to broker-request %s", "");
+        return 1;
+    }
+
+    if (!dbus_broker_policy) {
+        DBUS_BROKER_WARNING("No policy in place %s", "");
+        return 1;
+    }
+
+    int policy = 0;
+
+    struct etc_policy etc = dbus_broker_policy->etc;
+
+    int i, j;
+    for (i=0; i < etc.count; i++)
+        policy = filter(&(etc.rules[i]), dmsg, domid);
+
+    int domains = dbus_broker_policy->domain_number;
+
+    for (i=0; i < domains; i++) {
+        if (dbus_broker_policy->domains[i].domid == domid) {
+
+            struct domain_policy domain = dbus_broker_policy->domains[i];
+
+            for (j=0; j < domain.count; j++)
+                policy = filter(&(domain.rules[j]), dmsg, domid);
```

Comment:
Here and above, policy is stored on every filter() call.  Therefore only the last filter() result is returned from this function.  Either break on failure or |= the returns (provided success==0 and failure!=0).